Wikipedia/Code/AccountCreationViewController.swift

```diff
@@ -0,0 +1,285 @@
+               _captchaUrl = newValue
+                if (_captchaUrl) != nil {
+                    refreshCaptchaImage()
+                }
+            }
+        }
+    }
+    
+    fileprivate func refreshCaptchaImage() {
+        captchaViewController?.captchaTextBox.text = ""
+        captchaViewController?.captchaImageView.sd_setImage(with: captchaUrl)
+        showCaptchaContainer = true
+    }
+
+    func reloadCaptchaPushed(_ sender: AnyObject) {
+        captchaViewController?.captchaTextBox.text = ""
+        WMFAlertManager.sharedInstance.showAlert(localizedStringForKeyFallingBackOnEnglish("account-creation-captcha-obtaining"), sticky: false, dismissPreviousAlerts: true, tapCallBack: nil)
+        save() //this restarts the token process which gets us another captcha
+    }
+    
+    fileprivate func login() {
+        WMFAlertManager.sharedInstance.showAlert(localizedStringForKeyFallingBackOnEnglish("account-creation-logging-in"), sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
+        WMFAuthenticationManager.sharedInstance().login(
+            withUsername: usernameField.text!,
+            password: passwordField.text!,
+            success: {
+                let loggedInMessage = localizedStringForKeyFallingBackOnEnglish("main-menu-account-title-logged-in").replacingOccurrences(of: "$1", with: self.usernameField.text!)
+                WMFAlertManager.sharedInstance.showSuccessAlert(loggedInMessage, sticky: false, dismissPreviousAlerts: true, tapCallBack: nil)
+                self.dismiss(animated: true, completion: nil)
+        }, failure: { error in
+            self.enableProgressiveButton(true)
+            // WMFAlertManager.sharedInstance.showErrorAlert(error as NSError, sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
+        })
+    }
+    
+    fileprivate func requiredInputFields() -> [UITextField] {
+        assert(isViewLoaded, "This method is only intended to be called when view is loaded, since they'll all be nil otherwise")
+        return [usernameField, passwordField, passwordRepeatField]
+    }
+
+    fileprivate func isPasswordConfirmationCorrect() -> Bool {
+        return passwordField.text == passwordRepeatField.text
+    }
+    
+    fileprivate func areRequiredFieldsPopulated() -> Bool {
+        let firstRequiredFieldWithNoText = requiredInputFields().first(where:{ $0.text?.characters.count == 0 })
+        return firstRequiredFieldWithNoText == nil
+    }
+    
+    func doneButtonPushed(_ tap: UITapGestureRecognizer) {
+        WMFAlertManager.sharedInstance.showAlert(localizedStringForKeyFallingBackOnEnglish("account-creation-saving"), sticky: true, dismissPreviousAlerts: true, tapCallBack: nil)
+        save()
+    }
+    
+    override func updateViewConstraints() {
+        adjustScrollLimitForCaptchaVisiblity()
+        super.updateViewConstraints()
+    }
+
+    fileprivate func save() {
+        if !areRequiredFieldsPopulated() {
```

Comment:
These ifs should be guard statements so that the compiler enforces the early return